# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/angle/checkout/src/compiler/translator/tree_ops/EmulateGLDrawID.cpp
# Commit: 3a91885d47e0
# Full Hash: 3a91885d47e0ef7fba9749562d52300eea67c651
# Author: Miko Mynttinen <mikokm@gmail.com>
# Date: 2019-11-11 21:52:52
# Regressor Bug: 1578576
# File Overlap Count: 4
# Description:
#   Bug 1578576 - Part 2: Update to ANGLE 3865 r=jgilbert
#   
#   Depends on D44561
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D44579
# ==============================================================================

diff -r 77cee77c0307 -r 3a91885d47e0 gfx/angle/checkout/src/compiler/translator/tree_ops/EmulateGLDrawID.cpp
--- a/gfx/angle/checkout/src/compiler/translator/tree_ops/EmulateGLDrawID.cpp	Fri Nov 08 16:41:42 2019 +0000
+++ /dev/null	Thu Jan 01 00:00:00 1970 +0000
@@ -1,89 +0,0 @@
-//
-// Copyright 2018 The ANGLE Project Authors. All rights reserved.
-// Use of this source code is governed by a BSD-style license that can be
-// found in the LICENSE file.
-//
-// EmulateGLDrawID is an AST traverser to convert the gl_DrawID builtin
-// to a uniform int
-//
-
-#include "compiler/translator/tree_ops/EmulateGLDrawID.h"
-
-#include "angle_gl.h"
-#include "compiler/translator/StaticType.h"
-#include "compiler/translator/Symbol.h"
-#include "compiler/translator/SymbolTable.h"
-#include "compiler/translator/tree_util/BuiltIn_autogen.h"
-#include "compiler/translator/tree_util/IntermTraverse.h"
-#include "compiler/translator/tree_util/ReplaceVariable.h"
-#include "compiler/translator/util.h"
-
-namespace sh
-{
-
-namespace
-{
-
-constexpr const ImmutableString kEmulatedGLDrawIDName("angle_DrawID");
-
-class FindGLDrawIDTraverser : public TIntermTraverser
-{
-  public:
-    FindGLDrawIDTraverser() : TIntermTraverser(true, false, false), mVariable(nullptr) {}
-
-    const TVariable *getGLDrawIDBuiltinVariable() { return mVariable; }
-
-  protected:
-    void visitSymbol(TIntermSymbol *node) override
-    {
-        if (&node->variable() == BuiltInVariable::gl_DrawID() ||
-            &node->variable() == BuiltInVariable::gl_DrawIDESSL1())
-        {
-            mVariable = &node->variable();
-        }
-    }
-
-  private:
-    const TVariable *mVariable;
-};
-
-}  // namespace
-
-void EmulateGLDrawID(TIntermBlock *root,
-                     TSymbolTable *symbolTable,
-                     std::vector<sh::Uniform> *uniforms,
-                     bool shouldCollect)
-{
-    FindGLDrawIDTraverser traverser;
-    root->traverse(&traverser);
-    const TVariable *builtInVariable = traverser.getGLDrawIDBuiltinVariable();
-    if (builtInVariable)
-    {
-        const TType *type = StaticType::Get<EbtInt, EbpHigh, EvqUniform, 1, 1>();
-        const TVariable *drawID =
-            new TVariable(symbolTable, kEmulatedGLDrawIDName, type, SymbolType::AngleInternal);
-
-        // AngleInternal variables don't get collected
-        if (shouldCollect)
-        {
-            Uniform uniform;
-            uniform.name       = kEmulatedGLDrawIDName.data();
-            uniform.mappedName = kEmulatedGLDrawIDName.data();
-            uniform.type       = GLVariableType(*type);
-            uniform.precision  = GLVariablePrecision(*type);
-            uniform.staticUse  = symbolTable->isStaticallyUsed(*builtInVariable);
-            uniform.active     = true;
-            uniform.binding    = type->getLayoutQualifier().binding;
-            uniform.location   = type->getLayoutQualifier().location;
-            uniform.offset     = type->getLayoutQualifier().offset;
-            uniform.readonly   = type->getMemoryQualifier().readonly;
-            uniform.writeonly  = type->getMemoryQualifier().writeonly;
-            uniforms->push_back(uniform);
-        }
-
-        DeclareGlobalVariable(root, drawID);
-        ReplaceVariable(root, builtInVariable, drawID);
-    }
-}
-
-}  // namespace sh